File: semestre-2/disciplina-3-RENOMEAR/unidade-1/aula-3/aula-3.java
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.atomic.AtomicInteger;

/**
 * Representa um item no banco de dados.
 * Simula o dado persistente no Servidor/Nuvem.
 */
class Produto {
    private String nome;
    // Usamos AtomicInteger para simular a thread-safety (Isolamento/Consistência)
    // necessária em um ambiente de sistema distribuído (múltiplos acessos simultâneos).
    private AtomicInteger estoque; 

    public Produto(String nome, int estoqueInicial) {
        this.nome = nome;
        this.estoque = new AtomicInteger(estoqueInicial);
    }

    public String getNome() {
        return nome;
    }

    public int getEstoque() {
        return estoque.get();
    }

    /**
     * Tenta decrementar o estoque. 
     * Simula a transação e garante Atomicidade e Consistência (ACID).
     * @param quantidade A ser vendida.
     * @return true se a venda foi bem-sucedida (commit), false se falhou (rollback).
     */
    public boolean tentarVenda(int quantidade) {
        int estoqueAtual;
        int novoEstoque;
        
        // Loop de verificação otimista para garantir que 
        // nenhuma outra thread alterou o valor entre a leitura e a escrita.
        do {
            estoqueAtual = estoque.get(); // Leitura (Isolamento)
            if (estoqueAtual < quantidade) {
                return false; // Falha na Consistência
            }
            novoEstoque = estoqueAtual - quantidade;
        // O compareAndSet garante que a operação só será escrita se o estoqueAtual 
        // for o mesmo que lemos inicialmente, garantindo a Atomicidade.
        } while (!estoque.compareAndSet(estoqueAtual, novoEstoque));
        
        // Durabilidade: O valor foi persistido (no AtomicInteger, simulando o DB)
        return true; 
    }
}

/**
 * Simula o Servidor de Banco de Dados na Nuvem (Back-end).
 * Centraliza o gerenciamento de dados e a lógica transacional.
 */
class ServidorInventario {
    private Map<String, Produto> inventarioDB;

    public ServidorInventario() {
        // Simulação do Banco de Dados Central (Nuvem)
        inventarioDB = new HashMap<>();
        inventarioDB.put("SKU-100", new Produto("Camiseta Básica", 500));
        inventarioDB.put("SKU-201", new Produto("Calça Jeans Premium", 250));
    }

    /**
     * Simula a consulta de um Cliente. 
     * É a base da Arquitetura Cliente-Servidor.
     */
    public int consultarEstoque(String sku) {
        if (inventarioDB.containsKey(sku)) {
            return inventarioDB.get(sku).getEstoque();
        }
        return -1; // Produto não encontrado
    }

    /**
     * Processa o pedido de venda de um Cliente (Nó Distribuído).
     */
    public String realizarVenda(String sku, int quantidade, String nomeCliente) {
        System.out.println(String.format("[SERVIDOR]: Processando pedido de %s: Vender %d de %s.", nomeCliente, quantidade, sku));
        
        Produto produto = inventarioDB.get(sku);

        if (produto == null) {
            return "[SERVIDOR] ❌ Erro: SKU não existe.";
        }

        if (produto.tentarVenda(quantidade)) {
            // Sucesso na transação ACID.
            return String.format("[SERVIDOR] ✅ SUCESSO! %d un. vendidas. Novo Estoque Global: %d", 
                                 quantidade, produto.getEstoque());
        } else {
            // Falha na transação.
            return String.format("[SERVIDOR] ⛔ FALHA! Estoque insuficiente para %s. Atual: %d", 
                                 nomeCliente, produto.getEstoque());
        }
    }
}

/**
 * Simula um Cliente (Loja ou E-commerce) fazendo requisições ao Servidor.
 * Reforça o conceito de Sistema Distribuído (múltiplos pontos de acesso).
 */
class ClienteVarejo implements Runnable {
    private ServidorInventario servidor;
    private String nomeCliente;
    private String sku;
    private int quantidade;

    public ClienteVarejo(ServidorInventario servidor, String nomeCliente, String sku, int quantidade) {
        this.servidor = servidor;
        this.nomeCliente = nomeCliente;
        this.sku = sku;
        this.quantidade = quantidade;
    }

    @Override
    public void run() {
        // 1. Cliente consulta o estoque (Requisição GET)
        int estoqueAntes = servidor.consultarEstoque(sku);
        System.out.println(String.format("\n[CLIENTE %s]: Estoque antes: %d", nomeCliente, estoqueAntes));
        
        // Simula o tempo de processamento/comunicação
        try {
            Thread.sleep((long) (Math.random() * 100)); 
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }

        // 2. Cliente realiza a venda (Requisição POST/Transação)
        String resultado = servidor.realizarVenda(sku, quantidade, nomeCliente);
        System.out.println(String.format("[CLIENTE %s]: Resposta: %s", nomeCliente, resultado));
    }
}


public class aula_3 {
    public static void main(String[] args) throws InterruptedException {
        
        System.out.println("===================================================================");
        System.out.println("      SIMULAÇÃO JAVA: ARQUITETURA DISTRIBUÍDA EM NUVEM (THREADS)     ");
        System.out.println("===================================================================");
        
        ServidorInventario servidorCentral = new ServidorInventario();

        System.out.println("\n--- ESTOQUE INICIAL (SKU-100) ---");
        System.out.println("Estoque Total: " + servidorCentral.consultarEstoque("SKU-100"));

        // 1. Criando Clientes (Nós) que acessam o Servidor (Nuvem) simultaneamente (Threads)
        Thread t1 = new Thread(new ClienteVarejo(servidorCentral, "Loja-SP", "SKU-100", 200));
        Thread t2 = new Thread(new ClienteVarejo(servidorCentral, "E-commerce (Pico)", "SKU-100", 250));
        Thread t3 = new Thread(new ClienteVarejo(servidorCentral, "Loja-RJ", "SKU-100", 100)); // Deve falhar

        System.out.println("\n--- INICIANDO TRANSAÇÕES SIMULTÂNEAS (Sistema Distribuído) ---");
        
        // 2. Executando as transações
        t1.start();
        t2.start();
        t3.start();

        // Espera todas as threads terminarem para verificar o resultado final
        t1.join();
        t2.join();
        t3.join();

        // 3. Verificação final do estoque global sincronizado
        System.out.println("\n--- ESTOQUE FINAL SINCRONIZADO (Consistência) ---");
        int estoqueFinal = servidorCentral.consultarEstoque("SKU-100");
        System.out.println("Estoque Total Remanescente: " + estoqueFinal);
        
        // 500 (Inicial) - 200 (SP) - 250 (E-comm) = 50.
        // A transação de 100 da Loja-RJ deve falhar, demonstrando a Integridade (ACID).
        
        System.out.println("\n===================================================================");
        System.out.println(" O Java simula como o Servidor garante a integridade (AtomicInteger)");
        System.out.println("  e a sincronização de dados entre múltiplos Clientes/Nós.");
        System.out.println("===================================================================");
    }
}

File: semestre-2/disciplina-3-RENOMEAR/unidade-1/aula-3/aula-3.js
// O JavaScript será usado para simular a comunicação entre Cliente e Servidor
// na arquitetura de banco de dados distribuída em Nuvem.

// --- 1. O Servidor na Nuvem (Back-end) ---
// Simulado por um objeto em memória, representando o Banco de Dados Centralizado (Cloud DB)
const cloudInventoryDB = {
    'SKU-100': { nome: 'Camiseta Básica', estoque: 500, preco: 49.90 },
    'SKU-201': { nome: 'Calça Jeans Premium', estoque: 250, preco: 199.90 },
    'SKU-302': { nome: 'Tênis Esportivo', estoque: 800, preco: 249.90 },
};

/**
 * Simula o Servidor de Banco de Dados Cliente-Servidor.
 * Recebe a solicitação do Cliente e processa a lógica de negócio.
 * @param {string} sku - O ID do produto.
 * @returns {object} O status e os dados do inventário.
 */
function getServerInventory(sku) {
    console.log(`\n[SERVIDOR]: Recebida solicitação de inventário para SKU: ${sku}`);
    
    // Simula o tempo de processamento/consulta ao DB em Nuvem (latency)
    // new Promise(resolve => setTimeout(resolve, 100)); 

    const item = cloudInventoryDB[sku];
    
    if (item) {
        // Resposta do Servidor: Retorna o dataset (Estoque e Nome)
        return {
            status: 200,
            data: { 
                nome: item.nome, 
                estoque: item.estoque, 
                preco: item.preco 
            }
        };
    } else {
        // Resposta de Erro do Servidor
        return {
            status: 404,
            message: `Produto SKU ${sku} não encontrado.`
        };
    }
}

/**
 * Simula a transação de venda, garantindo a atomicidade e consistência.
 * @param {string} sku - O ID do produto.
 * @param {number} quantidade - Quantidade a ser vendida.
 * @param {string} cliente - Identificação do cliente (Loja/E-commerce).
 * @returns {object} Status da transação.
 */
function processServerTransaction(sku, quantidade, cliente) {
    console.log(`\n[SERVIDOR]: Recebida transação de ${cliente} para vender ${quantidade} de ${sku}`);
    
    const item = cloudInventoryDB[sku];
    
    if (!item) {
        return { status: 404, message: 'Produto não existe.' };
    }
    
    let novoEstoque = item.estoque;

    // 1. Verificação de Consistência e Isolamento (Regras de Negócio)
    if (novoEstoque >= quantidade) {
        // 2. Atomicidade: A transação é efetuada
        novoEstoque -= quantidade;
        
        // Simula a escrita/commit no Banco de Dados Central (Durabilidade)
        cloudInventoryDB[sku].estoque = novoEstoque;
        
        return { 
            status: 200, 
            message: `Venda de ${quantidade} de ${item.nome} efetuada por ${cliente}.`,
            novoEstoque: novoEstoque
        };
    } else {
        // Simula o Rollback implícito: A transação não é efetuada
        return { 
            status: 400, 
            message: `Estoque insuficiente para ${cliente}. Disponível: ${novoEstoque}.` 
        };
    }
}

// --- 2. Os Clientes (Front-ends) ---

/**
 * Simula o Cliente (Navegador/App PDV da Loja) fazendo uma requisição ao Servidor.
 * @param {string} sku - O ID do produto.
 * @param {string} nomeCliente - O nome do cliente (Loja-SP ou E-commerce).
 */
function clienteConsultaInventario(sku, nomeCliente) {
    console.log(`\n[CLIENTE ${nomeCliente}]: Requisição GET para verificar estoque do SKU ${sku}...`);
    
    const response = getServerInventory(sku);

    if (response.status === 200) {
        console.log(`[CLIENTE ${nomeCliente}]: ✅ Resposta Recebida.`);
        console.log(`   Produto: ${response.data.nome}`);
        console.log(`   Estoque Global Atual: ${response.data.estoque}`);
    } else {
        console.log(`[CLIENTE ${nomeCliente}]: ❌ Erro: ${response.message}`);
    }
    return response.data ? response.data.estoque : 0;
}

/**
 * Simula o Cliente fazendo um pedido de compra (Transação no sistema distribuído).
 */
function clienteRealizaCompra(sku, quantidade, nomeCliente) {
    console.log(`\n[CLIENTE ${nomeCliente}]: Requisição POST para comprar ${quantidade} de ${sku}...`);
    
    const response = processServerTransaction(sku, quantidade, nomeCliente);

    if (response.status === 200) {
        console.log(`[CLIENTE ${nomeCliente}]: ✅ Compra SUCESSO! ${response.message}`);
        console.log(`   Novo estoque no Servidor: ${response.novoEstoque}`);
    } else {
        console.log(`[CLIENTE ${nomeCliente}]: ⛔ Compra FALHOU: ${response.message}`);
    }
}

// ----------------- Fluxo de Simulação em JavaScript -----------------

console.log("========================================================================");
console.log("    SIMULAÇÃO JS: ARQUITETURA CLIENTE-SERVIDOR E DISTRIBUÍDA (NUVEM)    ");
console.log("========================================================================");

// 1. Cliente E-commerce consulta o estoque global antes de exibir a página.
const estoqueInicial = clienteConsultaInventario('SKU-201', 'E-commerce');

// 2. Clientes Distribuídos (Loja-SP e Loja-RJ) fazem vendas simultâneas.
// O Cliente/Loja faz a requisição, e o Servidor (Nuvel) garante a integridade.
clienteRealizaCompra('SKU-201', 5, 'Loja-SP - PDV');
clienteRealizaCompra('SKU-201', 20, 'E-commerce - Pedido');

// 3. Um terceiro Cliente (Loja-RJ) tenta uma grande venda após as anteriores.
// O servidor verifica o estoque atualizado (sincronizado em tempo real).
clienteRealizaCompra('SKU-201', 230, 'Loja-RJ - Venda Grande'); // Deve falhar, estoque em 225 (250 - 5 - 20)

// 4. E-commerce consulta o estoque novamente (Reflexo da sincronização).
clienteConsultaInventario('SKU-201', 'E-commerce');

console.log("\n========================================================================");
console.log("  O JS demonstra como o Cliente (Navegador/App) faz a Requisição e o");
console.log("  Servidor (Nuvel/Back-end) processa a Lógica e sincroniza o dado.");
console.log("========================================================================");